File: solutions/2026-07-11/python/2026-07-11-count-the-number-of-complete-components.py
```python
# Summary: Counts the number of complete connected components in a graph.
# A connected component is complete if every pair of vertices within it is connected by an edge.
# Link: https://leetcode.com/problems/count-the-number-of-complete-components/
#
# Approach:
# We can use Depth-First Search (DFS) or Breadth-First Search (BFS) to find connected components.
# For each connected component found, we then check if it's complete.
# To check if a component is complete:
# 1. Count the number of vertices (k) in the component.
# 2. Count the number of edges (m) within the component.
# 3. A component with k vertices is complete if and only if it has k * (k - 1) / 2 edges.
#
# We'll use an adjacency list to represent the graph. We'll also use a `visited` set to keep track
# of visited nodes during DFS/BFS.
#
# Time Complexity: O(N + E), where N is the number of vertices and E is the number of edges.
#   - Building the adjacency list takes O(N + E).
#   - Traversing each connected component using DFS/BFS takes O(N + E) in total across all components.
#   - For each component, checking completeness involves iterating through its vertices and edges.
#     The number of vertices and edges in all components sum up to N and E respectively.
#
# Space Complexity: O(N + E)
#   - Adjacency list: O(N + E)
#   - Visited set: O(N)
#   - Recursion stack for DFS (or queue for BFS): O(N) in the worst case (a single large component).

from collections import defaultdict, deque

class Solution:
    def countCompleteComponents(self, n: int, edges: list[list[int]]) -> int:
        # Build the adjacency list representation of the graph.
        adj = defaultdict(list)
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        # Keep track of visited vertices to avoid reprocessing components.
        visited = [False] * n
        complete_components_count = 0

        # Iterate through all vertices to find connected components.
        for i in range(n):
            if not visited[i]:
                # Found a new unvisited vertex, start DFS/BFS to find its component.
                component_nodes = []
                component_edges_count = 0
                queue = deque([i]) # Using BFS, could also use DFS (recursion or stack)
                visited[i] = True

                # Perform BFS to find all nodes in the current connected component.
                while queue:
                    node = queue.popleft()
                    component_nodes.append(node)

                    # Count the number of edges incident to this node *within the component*.
                    # Each edge will be counted twice (once for each endpoint), so we'll divide by 2 later.
                    component_edges_count += len(adj[node])

                    for neighbor in adj[node]:
                        if not visited[neighbor]:
                            visited[neighbor] = True
                            queue.append(neighbor)

                # Now, component_nodes contains all vertices of the current connected component.
                # component_edges_count is the sum of degrees of all nodes in the component.
                # The actual number of unique edges within the component is half of this sum.
                num_vertices_in_component = len(component_nodes)
                num_edges_in_component = component_edges_count // 2

                # Check if the component is complete.
                # A complete graph with k vertices has k * (k - 1) / 2 edges.
                expected_edges_for_complete_component = num_vertices_in_component * (num_vertices_in_component - 1) // 2

                if num_edges_in_component == expected_edges_for_complete_component:
                    complete_components_count += 1

        return complete_components_count

```

File: solutions/2026-07-11/javascript/2026-07-11-count-the-number-of-complete-components.js
// Problem: Count the Number of Complete Components
// Link: https://leetcode.com/problems/count-the-number-of-complete-components/
//
// Approach:
// We can use Depth First Search (DFS) or Breadth First Search (BFS) to find connected components.
// For each connected component, we need to verify if it's a "complete" component.
// A component is complete if every pair of vertices within that component has an edge.
//
// To do this efficiently:
// 1. Build an adjacency list representation of the graph.
// 2. Maintain a `visited` array to keep track of visited nodes.
// 3. Iterate through each vertex from 0 to n-1.
// 4. If a vertex hasn't been visited, start a DFS/BFS from it to find a connected component.
// 5. During the DFS/BFS traversal, collect all vertices belonging to the current component and count their degrees within that component.
// 6. After finding a component, check if it's complete. A component with `k` vertices is complete if each of its vertices has a degree of `k-1` *within that component*.
//    Also, ensure that the number of edges within the component is exactly `k * (k - 1) / 2`.
// 7. If the component is complete, increment a counter.
// 8. Return the final counter.
//
// Time Complexity:
// Building the adjacency list: O(E), where E is the number of edges.
// Traversing the graph: Each vertex and edge will be visited at most once across all DFS/BFS calls.
// For each component of size k and m edges, checking completeness takes O(k + m) (to count degrees and then compare with k*(k-1)/2).
// In the worst case, we might have a single large component.
// Total time complexity is roughly O(N + E), where N is the number of vertices and E is the number of edges.
// Given N <= 50, this is very efficient.
//
// Space Complexity:
// Adjacency list: O(N + E)
// Visited array: O(N)
// Recursion stack for DFS (or queue for BFS): O(N) in the worst case.
// Total space complexity is O(N + E).
// Given N <= 50, this is very efficient.

function countCompleteComponents(n, edges) {
    // 1. Build the adjacency list and count total edges for checking completeness later.
    const adj = new Array(n).fill(0).map(() => []);
    // Store the actual number of edges in the graph
    const totalEdges = edges.length;

    for (const [u, v] of edges) {
        adj[u].push(v);
        adj[v].push(u);
    }

    // Array to keep track of visited vertices.
    const visited = new Array(n).fill(false);
    // Counter for complete components.
    let completeComponentsCount = 0;

    // Iterate through each vertex to find connected components.
    for (let i = 0; i < n; i++) {
        // If the vertex hasn't been visited yet, it's the start of a new potential component.
        if (!visited[i]) {
            // Variables to store information about the current component being explored.
            let componentVertices = []; // List of vertices in the current component.
            let componentEdges = 0;     // Count of edges within the current component.

            // Start DFS to find all vertices and edges in this connected component.
            const stack = [i];
            visited[i] = true; // Mark the starting node as visited.

            // Perform DFS
            while (stack.length > 0) {
                const u = stack.pop();
                componentVertices.push(u); // Add vertex to the current component's list.

                // Count edges originating from this vertex that connect to other vertices *within this component*.
                // Each edge will be counted twice (once for each endpoint), so we'll divide by 2 later.
                componentEdges += adj[u].length;

                // Explore neighbors.
                for (const v of adj[u]) {
                    if (!visited[v]) {
                        visited[v] = true; // Mark neighbor as visited.
                        stack.push(v);     // Add neighbor to the stack for further exploration.
                    }
                }
            }

            // After DFS, we have all vertices (`componentVertices`) and the total degree sum (`componentEdges`).
            // The number of vertices in the component is `k`.
            const k = componentVertices.length;
            // The number of edges within this component is `m = componentEdges / 2`.
            const m = componentEdges / 2;

            // Check if the component is "complete".
            // A component is complete if:
            // 1. The number of vertices is `k`.
            // 2. The number of edges within this component is `k * (k - 1) / 2`.
            // This condition implies that every vertex has a degree of `k-1` within the component,
            // as the sum of degrees (`componentEdges`) must be `k * (k-1)`.
            if (m === (k * (k - 1)) / 2) {
                completeComponentsCount++; // If it's complete, increment the counter.
            }
        }
    }

    // Return the total count of complete components.
    return completeComponentsCount;
}
